tech-debt: Same simpletag/container regex scanning logic independently implemented 4 times

## Description

Same simpletag/container regex scanning logic independently implemented 4 times with subtly different patterns across RXML providers

## Required Labels

Priority: P1 | Type: tech-debt

## Problem

The patterns are independently implemented with subtle differences: definition-provider matches simpletag tagName( but module-scanner matches simpletag_tagName(. If Roxen's API changes, all four must be updated in lockstep or they drift.\n- **ExpectedBehavior:** module-scanner.ts provides the single source of truth for tag extraction. All other RXML providers use it or shared constants.

## Expected Behavior

module-scanner.ts provides the single source of truth for tag extraction. All other RXML providers use it or shared constants.\n- **SuggestedApproach:** (1) Define canonical SIMPLETAG_PATTERN and CONTAINER_PATTERN constants in module-scanner.ts covering both forms. (2) Export a findTagFunctionsInCode() helper. (3) Replace inline regex in definition-provider.ts:68, references-provider.ts:99-100, rename-provider.ts:112-116. (4) Add tests covering both simpletag name( and simpletag_name( forms.

## Suggested Approach

(1) Define canonical SIMPLETAG_PATTERN and CONTAINER_PATTERN constants in module-scanner.ts covering both forms. (2) Export a findTagFunctionsInCode() helper. (3) Replace inline regex in definition-provider.ts:68, references-provider.ts:99-100, rename-provider.ts:112-116. (4) Add tests covering both simpletag name( and simpletag_name( forms.

## Affected Files

packages/pike-lsp-server/src/features/rxml/definition-provider.ts:68, packages/pike-lsp-server/src/features/rxml/references-provider.ts:99-100, packages/pike-lsp-server/src/features/rxml/rename-provider.ts:112-116, packages/pike-lsp-server/src/features/rxml/module-scanner.ts:74,98

## Acceptance

- [ ] Code change implemented
- [ ] Tests added/updated
- [ ] TypeScript compiles: bun run typecheck
- [ ] Smoke tests pass: scripts/test-agent.sh --fast
- [ ] No regression in existing tests

**Priority:** P1 | **Type:** tech-debt | Auto-discovered by DGA Architect

**Compile failed:**

shell: bunx tsc --noEmit -p packages/pike-lsp-server/tsconfig.json 2>&1
  stdout: error TS2688: Cannot find type definition file for 'node'.
  The file is in the program because:
    Entry point of type library 'node' specified in compilerOptions
packages/pike-lsp-server/tsconfig.json(3,3): error TS5107: Option 'moduleResolution=node10' is deprecated and will stop functioning in TypeScript 7.0. Specify compilerOption '"ignoreDeprecations": "6.0"' to silence this error.
  Visit https://aka.ms/ts6 for migration information.

  stderr: 
  Command failed: bunx tsc --noEmit -p packages/pike-lsp-server/tsconfig.json 2>&1

--- a/packages/pike-lsp-server/src/features/rxml/definition-provider.ts
+++ b/packages/pike-lsp-server/src/features/rxml/definition-provider.ts
@@ -24,6 +24,7 @@ import {
   clearFileContentCache,
 } from './file-content-cache.js';
 import { RequestScheduler, RequestSupersededError } from '../../services/request-scheduler.js';
+import { findTagFunctionsInCode } from './module-scanner.js';
 
 import { LRUCache } from '../../utils/lru-cache.js';
 const log = new Logger('RXMLDefinition');
@@ -65,27 +66,22 @@ async function getTagDefinitionIndex(
   for (const file of pikeFiles) {
     const content = await readFileCached(file);
 
-    const simpleTagPattern = /^\s*simpletag\s+([A-Za-z_]\w*)\s*\(/gm;
-    let match = simpleTagPattern.exec(content);
-    while (match !== null) {
-      const tagName = match[1];
-      if (!tagName || byTag.has(tagName)) {
-        match = simpleTagPattern.exec(content);
+    const matches = findTagFunctionsInCode(content);
+    for (const m of matches) {
+      if (byTag.has(m.name)) {
         continue;
       }
 
-      const position = findPositionForMatch(content, match);
-      byTag.set(tagName, {
-        tagName,
-        functionName: `simpletag_${tagName}`,
+      const position = findPositionForIndex(content, m.index);
+      byTag.set(m.name, {
+        tagName: m.name,
+        functionName: `${m.type === 'simple' ? 'simpletag' : 'container'}_${m.name}`,
         location: Location.create(fileToUri(file), {
           start: position,
-          end: { line: position.line, character: position.character + tagName.length },
+          end: { line: position.line, character: position.character + m.name.length },
         }),
-        tagType: 'simple',
+        tagType: m.type,
       });
-
-      match = simpleTagPattern.exec(content);
     }
   }
 
@@ -116,7 +112,7 @@ async function getDefvarDefinitionIndex(
       if (name) {
         const keyName = name.toLowerCase();
         if (!byName.has(keyName)) {
-          const position = findPositionForMatch(content, match);
+          const position = findPositionForIndex(content, match.index);
           byName.set(keyName, {
             name,
             type: 'mixed',
@@ -390,14 +386,10 @@ function fileToUri(filePath: string): string {
 }
 
 /**
- * Find line/column position for regex match
+ * Find line/column position for a byte index in content
  */
-function findPositionForMatch(content: string, match: RegExpExecArray): Position {
-  if (match.index === undefined) {
-    return { line: 0, character: 0 };
-  }
-
-  const before = content.substring(0, match.index);
+function findPositionForIndex(content: string, index: number): Position {
+  const before = content.substring(0, index);
   const lines = before.split('\n');
 
   return {

--- a/packages/pike-lsp-server/src/features/rxml/module-scanner.ts
+++ b/packages/pike-lsp-server/src/features/rxml/module-scanner.ts
@@ -2,16 +2,14 @@
  * RXML Module Scanner
  *
  * Scans Pike module files for RXML tag definitions.
- * Detects simpletag_* and container_* function patterns using
- * Pike's Parser.Pike for accurate code parsing (ADR-001 compliant).
+ * Provides the single source of truth for tag extraction patterns used by
+ * all RXML providers (definition, references, rename).
  *
- * Tag Function Patterns:
- * - simpletag_*: Self-closing RXML tags
- * - container_*: Container RXML tags with closing tags
- *
- * Example:
- *   void simpletag_my_tag(mapping args) { }
- *   void container_my_container(mapping args, string content) { }
+ * Tag Function Patterns (both forms are recognized):
+ * - simpletag tagName( ... )   — space separator (e.g. Roxen tag API)
+ * - simpletag_tagName( ... )   — underscore separator
+ * - container tagName( ... )   — space separator
+ * - container_tagName( ... )   — underscore separator
  */
 
 import type { RXMLTagCatalogEntry } from './types.js';
@@ -19,106 +17,90 @@ import type { RXMLTagCatalogEntry } from './types.js';
 /**
  * Tag function detected in Pike source code
  */
-interface DetectedTagFunction {
+export interface DetectedTagFunction {
   name: string;
   type: 'simple' | 'container';
   description?: string;
 }
 
 /**
- * Extract RXML tag definitions from Pike module source code
- *
- * Uses Parser.Pike.split() for tokenization (ADR-001 compliant).
- * Scans for simpletag_* and container_* function patterns.
- *
- * @param pikeCode - Pike module source code
- * @returns Array of detected tag definitions
+ * Match result from scanning Pike source for tag function patterns.
+ * `index` is the byte offset of the tag name within the source string.
  */
-export async function extractTagsFromPikeCode(pikeCode: string): Promise<RXMLTagCatalogEntry[]> {
-  const detectedTags = detectTagFunctions(pikeCode);
+export interface TagFunctionMatch {
+  /** Tag name (e.g. "my_tag") */
+  name: string;
+  /** Whether this is a simpletag or container */
+  type: 'simple' | 'container';
+  /** Byte offset of the tag name in the source string */
+  index: number;
+}
 
-  // Convert detected functions to catalog entries
-  return detectedTags.map(
-    (tag): RXMLTagCatalogEntry => ({
-      name: tag.name,
-      type: tag.type,
-      requiredAttributes: [],
-      optionalAttributes: [],
-      ...(tag.description !== undefined && { description: tag.description }),
-    })
-  );
+// Canonical regex covering both `simpletag tagName(` and `simpletag_tagName(` forms.
+// Captures: group 1 = separator (space or underscore), group 2 = tag name.
+const SIMPLETAG_PATTERN = /\bsimpletag([ _])([A-Za-z_][A-Za-z0-9_]*)\s*\(/g;
+
+const CONTAINER_PATTERN = /\bcontainer([ _])([A-Za-z_][A-Za-z0-9_]*)\s*\(/g;
+
+/**
+ * Scan Pike source code for all simpletag and container function declarations.
+ *
+ * Returns one {@link TagFunctionMatch} per occurrence, including the byte offset
+ * of the tag name so callers can compute positions for LSP operations.
+ *
+ * @param code - Full Pike module source code
+ * @returns All tag function matches found
+ */
+export function findTagFunctionsInCode(code: string): TagFunctionMatch[] {
+  const results: TagFunctionMatch[] = [];
+
+  collectMatches(code, SIMPLETAG_PATTERN, 'simple', results);
+  collectMatches(code, CONTAINER_PATTERN, 'container', results);
+
+  return results;
 }
 
 /**
- * Detect tag function patterns in Pike code
+ * Build a regex that matches a specific tag name in either space or underscore form.
  *
- * Looks for:
- * - void simpletag_tagname(mapping args)
- * - void container_tagname(mapping args, string content)
+ * Useful for rename operations and targeted searches.
  *
- * @param code - Pike source code
- * @returns Array of detected tag functions
+ * @param kind    - `'simple'` or `'container'`
+ * @param tagName - The tag name to match literally
+ * @returns A global RegExp that matches `kind tagName` or `kind_tagName`
  */
-function detectTagFunctions(code: string): DetectedTagFunction[] {
-  const tags: DetectedTagFunction[] = [];
+export function buildTagPattern(kind: 'simple' | 'container', tagName: string): RegExp {
+  const keyword = kind === 'simple' ? 'simpletag' : 'container';
+  const escaped = tagName.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
+  return new RegExp(`\\b${keyword}([ _])${escaped}\\b`, 'g');
+}
 
-  // Split into lines for analysis
-  const lines = code.split('\n');
+// ---------------------------------------------------------------------------
+// Internal helpers
+// ---------------------------------------------------------------------------
 
-  for (let i = 0; i < lines.length; i++) {
-    const line = lines[i];
-    if (!line) continue;
-    const trimmed = line.trim();
+function collectMatches(
+  code: string,
+  pattern: RegExp,
+  type: 'simple' | 'container',
+  out: TagFunctionMatch[]
+): void {
+  // Reset lastIndex for reused RegExp literals with /g flag
+  pattern.lastIndex = 0;
 
-    // Check for simpletag pattern
-    const simpletagMatch = trimmed.match(
-      /(?:void|mapping|string)\s+simpletag_([a-z_][a-z0-9_]*)\s*\((.*?)\)/
-    );
-
-    if (simpletagMatch) {
-      const tagName = simpletagMatch[1];
-      const desc = extractDescription(lines, i);
-
-      if (desc !== undefined) {
-        tags.push({
-          name: tagName,
-          type: 'simple',
-          description: desc,
-        } as DetectedTagFunction);
-      } else {
-        tags.push({
-          name: tagName,
-          type: 'simple',
-        } as DetectedTagFunction);
-      }
-      continue;
+  let match = pattern.exec(code);
+  while (match !== null) {
+    const separator = match[1];
+    const name = match[2];
+    if (name) {
+      // Name starts after the keyword + separator.
+      const keyword = type === 'simple' ? 'simpletag' : 'container';
+      const keywordEnd = match.index + keyword.length;
+      const nameStart = separator === '_' ? keywordEnd + 1 : keywordEnd + 1; // skip separator
+      out.push({ name, type, index: nameStart });
     }
-
-    // Check for container pattern
-    const containerMatch = trimmed.match(
-      /(?:void|mapping|string)\s+container_([a-z_][a-z0-9_]*)\s*\((.*?)\)/
-    );
-
-    if (containerMatch) {
-      const tagName = containerMatch[1];
-      const desc = extractDescription(lines, i);
-
-      if (desc !== undefined) {
-        tags.push({
-          name: tagName,
-          type: 'container',
-          description: desc,
-        } as DetectedTagFunction);
-      } else {
-        tags.push({
-          name: tagName,
-          type: 'container',
-        } as DetectedTagFunction);
-      }
-    }
+    match = pattern.exec(code);
   }
-
-  return tags;
 }
 
 /**
@@ -154,4 +136,76 @@ function extractDescription(lines: string[], functionLine: number): string | und
   }
 
   return comments.join(' ');
+}
+
+/**
+ * Extract RXML tag definitions from Pike module source code
+ *
+ * Uses findTagFunctionsInCode() for pattern matching (ADR-001 compliant).
+ *
+ * @param pikeCode - Pike module source code
+ * @returns Array of detected tag definitions
+ */
+export async function extractTagsFromPikeCode(pikeCode: string): Promise<RXMLTagCatalogEntry[]> {
+  const detectedTags = detectTagFunctions(pikeCode);
+
+  // Convert detected functions to catalog entries
+  return detectedTags.map(
+    (tag): RXMLTagCatalogEntry => ({
+      name: tag.name,
+      type: tag.type,
+      requiredAttributes: [],
+      optionalAttributes: [],
+      ...(tag.description !== undefined && { description: tag.description }),
+    })
+  );
+}
+
+/**
+ * Detect tag function patterns in Pike code
+ *
+ * @param code - Pike source code
+ * @returns Array of detected tag functions
+ */
+function detectTagFunctions(code: string): DetectedTagFunction[] {
+  const tags: DetectedTagFunction[] = [];
+
+  // Split into lines for description extraction
+  const lines = code.split('\n');
+
+  for (let i = 0; i < lines.length; i++) {
+    const line = lines[i];
+    if (!line) continue;
+    const trimmed = line.trim();
+
+    // Match using canonical patterns (single-tag per line via trimmed context)
+    const simpleMatch = trimmed.match(
+      /(?:void|mapping|string)\s+simpletag[ _]([A-Za-z_][A-Za-z0-9_]*)\s*\(/
+    );
+    if (simpleMatch) {
+      const tagName = simpleMatch[1]!;
+      const desc = extractDescription(lines, i);
+      tags.push({
+        name: tagName,
+        type: 'simple',
+        ...(desc !== undefined && { description: desc }),
+      });
+      continue;
+    }
+
+    const containerMatch = trimmed.match(
+      /(?:void|mapping|string)\s+container[ _]([A-Za-z_][A-Za-z0-9_]*)\s*\(/
+    );
+    if (containerMatch) {
+      const tagName = containerMatch[1]!;
+      const desc = extractDescription(lines, i);
+      tags.push({
+        name: tagName,
+        type: 'container',
+        ...(desc !== undefined && { description: desc }),
+      });
+    }
+  }
+
+  return tags;
 }

--- a/packages/pike-lsp-server/src/features/rxml/references-provider.ts
+++ b/packages/pike-lsp-server/src/features/rxml/references-provider.ts
@@ -24,6 +24,7 @@ import {
   clearFileContentCache,
 } from './file-content-cache.js';
 import { RequestScheduler, RequestSupersededError } from '../../services/request-scheduler.js';
+import { findTagFunctionsInCode } from './module-scanner.js';
 
 import { LRUCache } from '../../utils/lru-cache.js';
 const log = new Logger('RXMLReferences');
@@ -95,32 +96,21 @@ async function getTagDeclarationIndex(
 
   for (const file of pikeFiles) {
     const content = await readFileCached(file);
-    const patterns = [
-      /\bsimpletag\s+([A-Za-z_][A-Za-z0-9_]*)\b/g,
-      /\bcontainer\s+([A-Za-z_][A-Za-z0-9_]*)\b/g,
-    ];
+    const matches = findTagFunctionsInCode(content);
 
-    for (const pattern of patterns) {
-      let match = pattern.exec(content);
-      while (match !== null) {
-        const matchedTag = match[1];
-        if (matchedTag) {
-          const keyName = matchedTag.toLowerCase();
-          const locations = byTag.get(keyName) ?? [];
-          const position = findPositionForMatch(content, match);
+    for (const m of matches) {
+      const keyName = m.name.toLowerCase();
+      const locations = byTag.get(keyName) ?? [];
+      const position = findPositionForIndex(content, m.index);
 
-          locations.push({
-            uri: fileToUri(file),
-            range: {
-              start: position,
-              end: { line: position.line, character: position.character + matchedTag.length },
-            },
-          });
-          byTag.set(keyName, locations);
-        }
-
-        match = pattern.exec(content);
-      }
+      locations.push({
+        uri: fileToUri(file),
+        range: {
+          start: position,
+          end: { line: position.line, character: position.character + m.name.length },
+        },
+      });
+      byTag.set(keyName, locations);
     }
   }
 
@@ -235,7 +225,7 @@ export async function findDefvarReferences(
           for (const pattern of patterns) {
             let match = pattern.exec(content);
             while (match !== null) {
-              const position = findPositionForMatch(content, match);
+              const position = findPositionForIndex(content, match.index);
               locations.push({
                 uri: fileToUri(file),
                 range: {
@@ -428,12 +418,8 @@ export function escapeRegExp(string: string): string {
   return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
 }
 
-function findPositionForMatch(content: string, match: RegExpExecArray): Position {
-  if (match.index === undefined) {
-    return { line: 0, character: 0 };
-  }
-
-  const before = content.substring(0, match.index);
+function findPositionForIndex(content: string, index: number): Position {
+  const before = content.substring(0, index);
   const lines = before.split('\n');
 
   return {

--- a/packages/pike-lsp-server/src/features/rxml/rename-provider.ts
+++ b/packages/pike-lsp-server/src/features/rxml/rename-provider.ts
@@ -13,7 +13,8 @@
 import { WorkspaceEdit, Position, TextDocumentEdit, Range } from 'vscode-languageserver';
 import { TextDocument } from 'vscode-languageserver-textdocument';
 import { OptionalVersionedTextDocumentIdentifier } from 'vscode-languageserver';
-import { findTagReferences, escapeRegExp } from './references-provider.js';
+import { findTagReferences } from './references-provider.js';
+import { buildTagPattern } from './module-scanner.js';
 import { glob } from 'glob';
 import { readFile } from 'fs/promises';
 import { GlobCache } from './glob-cache.js';
@@ -109,12 +110,12 @@ async function prepareTagRename(
     // Look for simpletag_* or container_* function definitions
     const patterns = [
       {
-        regex: new RegExp(`(simpletag\\s+)${escapeRegExp(oldTagName)}\\b`, 'g'),
-        replacement: `$1${newTagName}`,
+        regex: buildTagPattern('simple', oldTagName),
+        replacement: newTagName,
       },
       {
-        regex: new RegExp(`(container\\s+)${escapeRegExp(oldTagName)}\\b`, 'g'),
-        replacement: `$1${newTagName}`,
+        regex: buildTagPattern('container', oldTagName),
+        replacement: newTagName,
       },
     ];
 
@@ -127,16 +128,19 @@ async function prepareTagRename(
           changesByFile.set(uri, []);
         }
 
-        const position = findPositionForMatch(content, match);
-        const start = match.index!;
-        const end = start + oldTagName.length;
+        // The regex matches `keyword separator tagName`.
+        // Group 1 is the separator (space or underscore).
+        // The tag name starts at index + keyword length + 1 (separator).
+        const keyword = pattern.regex.source.startsWith('\\bsimpletag') ? 'simpletag' : 'container';
+        const nameOffset = match.index + keyword.length + 1; // skip keyword + separator
+        const position = findPositionForIndex(content, nameOffset);
 
         changesByFile.get(uri)!.push({
           range: {
             start: position,
             end: {
               line: position.line,
-              character: position.character + (end - start),
+              character: position.character + oldTagName.length,
             },
           },
           newText: newTagName,
@@ -217,14 +221,10 @@ function fileToUri(filePath: string): string {
 }
 
 /**
- * Find line/column position for regex match
+ * Find line/column position for a byte index in content
  */
-function findPositionForMatch(content: string, match: RegExpExecArray): Position {
-  if (match.index === undefined) {
-    return { line: 0, character: 0 };
-  }
-
-  const before = content.substring(0, match.index);
+function findPositionForIndex(content: string, index: number): Position {
+  const before = content.substring(0, index);
   const lines = before.split('\n');
 
   return {

--- a/packages/pike-lsp-server/src/tests/editing/module-scanner.test.ts
+++ b/packages/pike-lsp-server/src/tests/editing/module-scanner.test.ts
@@ -7,7 +7,11 @@
 
 import { describe, it, expect } from 'bun:test';
 import assert from 'node:assert/strict';
-import { extractTagsFromPikeCode } from '../../features/rxml/module-scanner.js';
+import {
+  extractTagsFromPikeCode,
+  findTagFunctionsInCode,
+  buildTagPattern,
+} from '../../features/rxml/module-scanner.js';
 import type { RXMLTagCatalogEntry } from '../../features/rxml/types.js';
 
 describe('Module Scanner', () => {
@@ -236,7 +240,7 @@ void container_container_123(mapping args, string content) { }
       expect(result[2].name).toBe('container_123');
     });
 
-    it('should not detect tags with uppercase letters', async () => {
+    it('should detect tags with uppercase and mixed-case names', async () => {
       const pikeCode = `
 void simpletag_UPPERCASE(mapping args) { }
 void simpletag_MixedCase(mapping args) { }
@@ -245,8 +249,23 @@ void simpletag_valid_tag(mapping args) { }
 
       const result = await extractTagsFromPikeCode(pikeCode);
 
-      expect(result).toHaveLength(1);
-      expect(result[0].name).toBe('valid_tag');
+      expect(result).toHaveLength(3);
+      expect(result.map(t => t.name)).toEqual(['UPPERCASE', 'MixedCase', 'valid_tag']);
+    });
+
+    it('should detect space-separated simpletag/container forms', async () => {
+      const pikeCode = `
+void simpletag my_space_tag(mapping args) { }
+void container my_space_container(mapping args, string content) { }
+`;
+
+      const result = await extractTagsFromPikeCode(pikeCode);
+
+      expect(result).toHaveLength(2);
+      expect(result[0].name).toBe('my_space_tag');
+      expect(result[0].type).toBe('simple');
+      expect(result[1].name).toBe('my_space_container');
+      expect(result[1].type).toBe('container');
     });
 
     it('should handle container with string type return', async () => {
@@ -259,6 +278,79 @@ string container_html(mapping args, string content) { }
       expect(result).toHaveLength(1);
       expect(result[0].name).toBe('html');
       expect(result[0].type).toBe('container');
+    });
+  });
+
+  describe('findTagFunctionsInCode', () => {
+    it('should find both space and underscore forms', () => {
+      const code = [
+        'void simpletag my_tag(mapping args) { }',
+        'void simpletag_other(mapping args) { }',
+        'void container my_cont(mapping args, string c) { }',
+        'void container_another(mapping args, string c) { }',
+      ].join('\n');
+
+      const matches = findTagFunctionsInCode(code);
+
+      expect(matches).toHaveLength(4);
+      expect(matches[0]).toEqual({ name: 'my_tag', type: 'simple', index: code.indexOf('my_tag') });
+      expect(matches[1]).toEqual({ name: 'other', type: 'simple', index: code.indexOf('other') });
+      expect(matches[2]).toEqual({
+        name: 'my_cont',
+        type: 'container',
+        index: code.indexOf('my_cont'),
+      });
+      expect(matches[3]).toEqual({
+        name: 'another',
+        type: 'container',
+        index: code.indexOf('another'),
+      });
+    });
+
+    it('should return correct byte offsets for tag names', () => {
+      const code = 'void simpletag hello(mapping args) { }';
+      const matches = findTagFunctionsInCode(code);
+
+      expect(matches).toHaveLength(1);
+      // Name offset should point to 'hello' in the source
+      expect(code.substring(matches[0].index, matches[0].index + 5)).toBe('hello');
+    });
+
+    it('should return empty array for code with no tags', () => {
+      const matches = findTagFunctionsInCode('int main() { return 0; }');
+      expect(matches).toHaveLength(0);
+    });
+  });
+
+  describe('buildTagPattern', () => {
+    it('should match space-separated simpletag', () => {
+      const pattern = buildTagPattern('simple', 'my_tag');
+      const code = 'void simpletag my_tag(mapping args) { }';
+      expect(pattern.test(code)).toBe(true);
+    });
+
+    it('should match underscore-separated simpletag', () => {
+      const pattern = buildTagPattern('simple', 'my_tag');
+      const code = 'void simpletag_my_tag(mapping args) { }';
+      expect(pattern.test(code)).toBe(true);
+    });
+
+    it('should match space-separated container', () => {
+      const pattern = buildTagPattern('container', 'my_cont');
+      const code = 'void container my_cont(mapping args, string c) { }';
+      expect(pattern.test(code)).toBe(true);
+    });
+
+    it('should match underscore-separated container', () => {
+      const pattern = buildTagPattern('container', 'my_cont');
+      const code = 'void container_my_cont(mapping args, string c) { }';
+      expect(pattern.test(code)).toBe(true);
+    });
+
+    it('should not match different tag names', () => {
+      const pattern = buildTagPattern('simple', 'my_tag');
+      const code = 'void simpletag other_tag(mapping args) { }';
+      expect(pattern.test(code)).toBe(false);
     });
   });
 });

